pkg/utils/kubernetes/utils.go

```diff
@@ -0,0 +1,130 @@
+package kubernetes
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"strings"
+)
+
+//OnFinalize Callback function to be executed when an object is being deleted
+type OnFinalize func() error
+
+type ObjectFinalizer struct {
+	object     runtime.Object
+	finalizers map[string]OnFinalize
+}
+
+type FinalizerManager struct {
+	objects map[types.UID]ObjectFinalizer
+	Client  client.Client
+}
+
+func NewFinalizerManager(client client.Client) FinalizerManager {
+	return FinalizerManager{
+		objects: map[types.UID]ObjectFinalizer{},
+		Client:  client,
+	}
+}
+
+//RegisterFinalizer registers a finalizer function to be executed when the object is marked to be deleted.
+func (mgr *FinalizerManager) RegisterFinalizer(owner metav1.Object, name string, onFinalize OnFinalize) error {
```

Comment:
I get the logic of this, but would really like to see how it fits in the one actual use case we have. Can you do a draft PR for your console link implementation?

My main concern is that the controller gets BC in a reconcile loop, and is likely to call RegisterFinalizer() at roughly the same spot in the code as it would call Finalize()

Reply:
Finalizers are usually added when the resource is reconciled and the Finalize() method should be at the beginning of the reconcile. Here is the example: https://github.com/ruromero/kie-cloud-operator/commit/5c280fa77a8f4fc531efe96e7d2532cb723062e8#diff-9ecb910ca7304714ab23ba15d8cf55a5